pkg/replication/replicator.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2021 CodeNotary, Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package replication
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/codenotary/immudb/pkg/api/schema"
+	"github.com/codenotary/immudb/pkg/client"
+	"github.com/codenotary/immudb/pkg/database"
+	"github.com/codenotary/immudb/pkg/logger"
+	"github.com/codenotary/immudb/pkg/stream"
+	"google.golang.org/grpc/metadata"
+)
+
+var ErrIllegalArguments = errors.New("illegal arguments")
+var ErrAlreadyRunning = errors.New("already running")
+var ErrAlreadyStopped = errors.New("already stopped")
+
+type TxReplicator struct {
+	db   database.DB
+	opts *Options
+
+	logger logger.Logger
+
+	mainContext context.Context
+	cancelFunc  context.CancelFunc
+
+	streamSrvFactory stream.ServiceFactory
+	client           client.ImmuClient
+	clientContext    context.Context
+
+	delayer        Delayer
+	failedAttempts int
+
+	nextTx uint64
+
+	running bool
+
+	mutex sync.Mutex
+}
+
+func NewTxReplicator(db database.DB, opts *Options, logger logger.Logger) (*TxReplicator, error) {
+	if db == nil || opts == nil || !opts.Valid() {
+		return nil, ErrIllegalArguments
+	}
+
+	return &TxReplicator{
+		db:               db,
+		opts:             opts,
+		logger:           logger,
+		streamSrvFactory: stream.NewStreamServiceFactory(opts.streamChunkSize),
+		delayer:          opts.delayer,
+	}, nil
+}
+
+func (txr *TxReplicator) Start() error {
+	txr.mutex.Lock()
+	defer txr.mutex.Unlock()
+
+	if txr.running {
+		return ErrAlreadyRunning
+	}
+
+	masterDB := fullAddress(txr.opts.masterDatabase, txr.opts.masterAddress, txr.opts.masterPort)
+
+	txr.logger.Infof("Initializing replication from '%s' to '%s'...", masterDB, txr.db.GetName())
+
+	st, err := txr.db.CurrentState()
+	if err != nil {
+		return err
+	}
+
+	txr.mainContext, txr.cancelFunc = context.WithCancel(context.Background())
+
+	txr.nextTx = st.TxId + 1
+
+	txr.running = true
+
+	go func() {
+		for {
```

Comment:
Shouldn't we also check if the context was not cancelled?

```suggestion
		for txr.mainContext.Err() == nil {
```

I smell some subtle timing issues that could cause reopening of the client connection after the `Stop` was called

Reply:
A corner case is if there's some connect / fetchTx error and both the timeout and context cancellation fire at the same time. Then even though context is cancelled, the second path (timeout) can be selected and will cause one more reconnect.

A very rare case thus I don't think it should block this PR.